# hypercarge_locations 

In [134]:
# Importing essential libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [135]:
# Load the dataset
hypercarge_location = pd.read_csv(r'dataset\full\hypercarge_locations.csv')

# Preview the first few rows
hypercarge_location.head()

chargerId serialNumber          sshurl  sshPort     gpsLat    gpsLong  \
0      20800    23BZ1509B   10.245.161.30      NaN  45.041390  11.710673   
1      17737    22BZ5330B   10.245.22.116      NaN  42.029548  11.962168   
2      83776    24BZ1716B  10.246.181.216      NaN  43.812700  11.154000   
3      19519    23BZ1033B   10.245.29.175      NaN  42.034378  12.648893   
4      83781    24BZ1721B   10.246.186.62      NaN        NaN        NaN   

   numberStacks  chassis  emergencyStop  doorContactSwitch  ...  \
0             4  HYC_300          False               True  ...   
1             4  HYC_300          False               True  ...   
2             4  HYC_300          False               True  ...   
3             4  HYC_300          False               True  ...   
4             4  HYC_300          False               True  ...   

    corporationName  operatorName  hasVollmacht hasCommissioningProtocol  \
0  Free To X S.p.A.           NaN         False                    False   
1  Free To X S.p.A.           NaN         False                    False   
2  Free To X S.p.A.           NaN         False                    False   
3  Free To X S.p.A.           NaN         False                    False   
4  Free To X S.p.A.           NaN         False                    False   

                                 surroundingChargers isFavourite  \
0  [{'chargerId': 6772, 'gpsLat': 44.66244, 'gpsL...       False   
1  [{'chargerId': 6771, 'gpsLat': 42.49917, 'gpsL...       False   
2  [{'chargerId': 6772, 'gpsLat': 44.66244, 'gpsL...       False   
3  [{'chargerId': 6771, 'gpsLat': 42.49917, 'gpsL...       False   
4                                                NaN       False   

   sessionLiveViewEnabled  finishedEol lendeeId lendeeName  
0                   False         True      NaN        NaN  
1                   False         True      NaN        NaN  
2                     NaN         True      NaN        NaN  
3                   False         True      NaN        NaN  
4                     NaN         True      NaN        NaN  

[5 rows x 74 columns]

In [136]:
# Check the statistics of numeric columns
hypercarge_location.describe()

chargerId  sshPort      gpsLat     gpsLong  numberStacks  \
count    374.000000      0.0  259.000000  259.000000    374.000000   
mean   43247.780749      NaN   41.117911    6.143900      3.823529   
std    33380.514195      NaN    9.314472   15.581104      0.421770   
min     6590.000000      NaN    0.000000  -50.000000      1.000000   
25%    13418.250000      NaN   41.849835    9.046675      4.000000   
50%    19747.500000      NaN   43.818792   11.419845      4.000000   
75%    83748.750000      NaN   44.962939   12.903881      4.000000   
max    83842.000000      NaN   57.000000   16.777855      4.000000   

           simIccid  serviceExpirationDate  softwareExpirationDate  \
count  3.720000e+02                    0.0                     0.0   
mean   8.945020e+18                    NaN                     NaN   
std    7.217696e+06                    NaN                     NaN   
min    8.945020e+18                    NaN                     NaN   
25%    8.945020e+18                    NaN                     NaN   
50%    8.945020e+18                    NaN                     NaN   
75%    8.945020e+18                    NaN                     NaN   
max    8.945020e+18                    NaN                     NaN   

       locationZipCode  calibrationLawOrdered  ...  defaultSoftwareVersion  \
count       153.000000                    0.0  ...                     0.0   
mean      35724.941176                    NaN  ...                     NaN   
std       24216.932744                    NaN  ...                     NaN   
min          10.000000                    NaN  ...                     NaN   
25%       20045.000000                    NaN  ...                     NaN   
50%       31032.000000                    NaN  ...                     NaN   
75%       50019.000000                    NaN  ...                     NaN   
max       83100.000000                    NaN  ...                     NaN   

       officialEichrechtCompletion  endClientId  distributorId  corporationId  \
count                          0.0          0.0     374.000000     374.000000   
mean                           NaN          NaN     184.350267     186.018717   
std                            NaN          NaN      84.130167      85.055996   
min                            NaN          NaN     180.000000     180.000000   
25%                            NaN          NaN     180.000000     180.000000   
50%                            NaN          NaN     180.000000     180.000000   
75%                            NaN          NaN     180.000000     180.000000   
max                            NaN          NaN    1807.000000    1807.000000   

       operatorId  endClientName  operatorName  lendeeId  lendeeName  
count         0.0            0.0           0.0       0.0         0.0  
mean          NaN            NaN           NaN       NaN         NaN  
std           NaN            NaN           NaN       NaN         NaN  
min           NaN            NaN           NaN       NaN         NaN  
25%           NaN            NaN           NaN       NaN         NaN  
50%           NaN            NaN           NaN       NaN         NaN  
75%           NaN            NaN           NaN       NaN         NaN  
max           NaN            NaN           NaN       NaN         NaN  

[8 rows x 24 columns]

## Handling missing values

In [137]:
# Check for missing values
hypercarge_location.isnull().sum()

chargerId                   0
serialNumber                0
sshurl                      0
sshPort                   374
gpsLat                    115
                         ... 
isFavourite                 0
sessionLiveViewEnabled    160
finishedEol                 0
lendeeId                  374
lendeeName                374
Length: 74, dtype: int64

In [138]:
# Check for duplicates
hypercarge_location.duplicated().sum()

0

The maximumn number of columns is 374, so I deleted all the features with 374 missing values.

In [139]:
# Check for columns with exactly 374 missing values
missing_value_count = hypercarge_location.isnull().sum()

# Identify columns where the missing value count is exactly 374
columns_to_drop = missing_value_count[missing_value_count == 374].index

# Drop those columns
hypercarge_location_cleaned = hypercarge_location.drop(columns=columns_to_drop)

# Check the result
print(f"Dropped columns: {list(columns_to_drop)}")
print(hypercarge_location_cleaned.head())

Dropped columns: ['sshPort', 'serviceExpirationDate', 'softwareExpirationDate', 'calibrationLawOrdered', 'lastMaintenance', 'defaultSoftwareVersion', 'officialEichrechtCompletion', 'endClientId', 'operatorId', 'endClientName', 'operatorName', 'lendeeId', 'lendeeName']
   chargerId serialNumber          sshurl     gpsLat    gpsLong  numberStacks  \
0      20800    23BZ1509B   10.245.161.30  45.041390  11.710673             4   
1      17737    22BZ5330B   10.245.22.116  42.029548  11.962168             4   
2      83776    24BZ1716B  10.246.181.216  43.812700  11.154000             4   
3      19519    23BZ1033B   10.245.29.175  42.034378  12.648893             4   
4      83781    24BZ1721B   10.246.186.62        NaN        NaN             4   

   chassis  emergencyStop  doorContactSwitch standAloneOrBackend  ...  \
0  HYC_300          False               True             Backend  ...   
1  HYC_300          False               True             Backend  ...   
2  HYC_300          False

New dataset name "hypercarge_location_cleaned" after deleting all features have 374 missing values.

In [140]:
# Find features with missing values
missing_values = hypercarge_location_cleaned.isnull().sum()

# Filter to show only columns with missing values
missing_values = missing_values[missing_values > 0]

# Display features with missing values and the count
print(missing_values)

gpsLat                    115
gpsLong                   115
simIccid                    2
commissioningDate         107
chargePointIdentity       113
customerIccid             336
warrantyExpiration        160
websocketUrl              114
locationStreet            221
locationTown              221
locationZipCode           221
locationProvince          221
locationCountry           147
locationUpdateNote        252
status_hycErrorCode        85
status_text               213
simActivationDate          35
simLastRequest             35
hymaintVersion             48
surroundingChargers       115
sessionLiveViewEnabled    160
dtype: int64


There are still a lot of features with too many missing values. Continue dropping features with > 200 missing values.

In [141]:
# Identify columns where the missing value count is exactly 374
columns_to_drop = missing_value_count[missing_value_count > 200].index

# Drop those columns
hypercarge_location_cleaned = hypercarge_location.drop(columns=columns_to_drop)

# Check the result
print(f"Dropped columns: {list(columns_to_drop)}")
print(hypercarge_location_cleaned.head())

Dropped columns: ['sshPort', 'customerIccid', 'serviceExpirationDate', 'softwareExpirationDate', 'locationStreet', 'locationTown', 'locationZipCode', 'locationProvince', 'locationUpdateNote', 'calibrationLawOrdered', 'status_text', 'lastMaintenance', 'defaultSoftwareVersion', 'officialEichrechtCompletion', 'endClientId', 'operatorId', 'endClientName', 'operatorName', 'lendeeId', 'lendeeName']
   chargerId serialNumber          sshurl     gpsLat    gpsLong  numberStacks  \
0      20800    23BZ1509B   10.245.161.30  45.041390  11.710673             4   
1      17737    22BZ5330B   10.245.22.116  42.029548  11.962168             4   
2      83776    24BZ1716B  10.246.181.216  43.812700  11.154000             4   
3      19519    23BZ1033B   10.245.29.175  42.034378  12.648893             4   
4      83781    24BZ1721B   10.246.186.62        NaN        NaN             4   

   chassis  emergencyStop  doorContactSwitch standAloneOrBackend  ...  \
0  HYC_300          False               True

In [142]:
# Find features with missing values
missing_values = hypercarge_location_cleaned.isnull().sum()

# Filter to show only columns with missing values
missing_values = missing_values[missing_values > 0]

# Display features with missing values and the count
print(missing_values)

gpsLat                    115
gpsLong                   115
simIccid                    2
commissioningDate         107
chargePointIdentity       113
warrantyExpiration        160
websocketUrl              114
locationCountry           147
status_hycErrorCode        85
simActivationDate          35
simLastRequest             35
hymaintVersion             48
surroundingChargers       115
sessionLiveViewEnabled    160
dtype: int64


simActivationDate and simLastRequest both have 35 missing values, I want to see if when simActivationDate = NaN, simLastRequest is also NaN

In [143]:
# Check if simLastRequest is NaN when simActivationDate is NaN
nan_check = hypercarge_location_cleaned[hypercarge_location_cleaned['simActivationDate'].isnull()]['simLastRequest'].isnull().value_counts()

# Display the results
print(nan_check)

simLastRequest
True    35
Name: count, dtype: int64


There is a feature simActivated (Boolean), I want to check if simActivated = False, both simActivationDate and simLastRequest = NaN

In [144]:
# Check if both simActivationDate and simLastRequest are NaN when simActivated is False
nan_check_sim_activated_false = hypercarge_location_cleaned.loc[
    hypercarge_location_cleaned['simActivated'] == False
]

# Count how many of those rows have simActivationDate and simLastRequest as NaN
count_nan_rows = nan_check_sim_activated_false[
    nan_check_sim_activated_false['simActivationDate'].isnull() & 
    nan_check_sim_activated_false['simLastRequest'].isnull()
].shape[0]

# Display the result
print("Number of rows where simActivated is False and both simActivationDate and simLastRequest are NaN:", count_nan_rows)

Number of rows where simActivated is False and both simActivationDate and simLastRequest are NaN: 35


Since both simActivationDate and simLastRequest are NaN when simActivated is False, their utility in the analysis may be limited => Drop these 2 features.

In [145]:
# Delete the features from the dataset
hypercarge_location_cleaned.drop(columns=['simActivationDate', 'simLastRequest'], inplace=True)

I wanted to check if simIccid important feature. First, let's see how many different values are there in that feature

In [146]:
# Count the number of unique values in the ICCID feature (simIccid)
unique_iccid_values_count = hypercarge_location_cleaned['simIccid'].nunique()

# Display the count of unique values
print("Number of unique values in simIccid:", unique_iccid_values_count)

Number of unique values in simIccid: 107


There are 107 different values which is a large number. There might be ok to delete this feature.

In [147]:
# Delete the IsimIccidCCID feature from the dataset
hypercarge_location_cleaned.drop(columns=['simIccid'], inplace=True)

Let's see if chargePointIdentity is worth to analysis.

In [148]:
# Count the number of unique values in the feature "Unique identity or name of the charge point"
unique_charge_point_identity_count = hypercarge_location_cleaned['chargePointIdentity'].nunique()

# Display the count of unique values
print("Number of unique values in chargePointIdentity:", unique_charge_point_identity_count)

Number of unique values in chargePointIdentity: 258


There are also too many different ID for this feature. Also, we already have ID for charger, I think it is not important to include this feature into analysing.

In [149]:
# Delete the chargePointIdentity feature from the dataset
hypercarge_location_cleaned.drop(columns=['chargePointIdentity'], inplace=True)

With too many missing values, warrantyExpiration (160), websocketUrl (114); hymaintVersion; commissioningDate (107) affect operational and maintenance aspects rather than directly influencing customer behaviour. It is reasonable to delete this feature.

In [150]:
# Delete warrantyExpiration, websocketUrl and hymaintVersion feature from the dataset
hypercarge_location_cleaned.drop(columns=['warrantyExpiration', 'websocketUrl', 'hymaintVersion', 'commissioningDate'], inplace=True)

Diagnose status_hycErrorCode

In [151]:
# Count the number of unique values in the feature "status_hycErrorCode" and display those values
unique_hyc_error_codes = hypercarge_location_cleaned['status_hycErrorCode'].value_counts()

# Display the count of unique values and the unique values themselves
unique_hyc_error_codes_count = unique_hyc_error_codes.shape[0]

print("Number of unique values in status_hycErrorCode:", unique_hyc_error_codes_count)
print("Unique values and their counts:\n", unique_hyc_error_codes)

Number of unique values in status_hycErrorCode: 2
Unique values and their counts:
 status_hycErrorCode
NoError        283
Door Opened      6
Name: count, dtype: int64


In [152]:
# Fill missing values in the status_hycErrorCode with "Unknown"
hypercarge_location_cleaned['status_hycErrorCode'] = hypercarge_location_cleaned['status_hycErrorCode'].fillna('Unknown')

Diagnose sessionLiveViewEnabled

In [153]:
# Count the number of unique values in the feature "status_hycErrorCode" and display those values
unique_sessionLiveViewEnabled = hypercarge_location_cleaned['sessionLiveViewEnabled'].value_counts()

# Display the count of unique values and the unique values themselves
unique_sessionLiveViewEnabled_count = unique_sessionLiveViewEnabled.shape[0]

print("Number of unique values in sessionLiveViewEnabled:", unique_sessionLiveViewEnabled_count)
print("Unique values and their counts:\n", unique_sessionLiveViewEnabled)

Number of unique values in sessionLiveViewEnabled: 1
Unique values and their counts:
 sessionLiveViewEnabled
False    214
Name: count, dtype: int64


There could be reasonable to consider missing value as True in this context

In [154]:
# Ensure the sessionLiveViewEnabled column is of type boolean before filling NaNs
hypercarge_location_cleaned['sessionLiveViewEnabled'] = hypercarge_location_cleaned['sessionLiveViewEnabled'].astype('boolean')

# Fill missing values in the sessionLiveViewEnabled with True
hypercarge_location_cleaned['sessionLiveViewEnabled'] = hypercarge_location_cleaned['sessionLiveViewEnabled'].fillna(True)

Since FreeX currently only operates in Italy => safely delete locationCountry

In [155]:
hypercarge_location_cleaned.drop(columns=['locationCountry'], inplace=True)

Diagnose gpsLat gpsLong and surroundingChargers:
They all have 115 missing values

In [156]:
# Check if gpsLat, gpsLong, and surroundingChargers all have NaN values simultaneously
missing_combination = hypercarge_location_cleaned[
    hypercarge_location_cleaned['gpsLat'].isnull() & 
    hypercarge_location_cleaned['gpsLong'].isnull() & 
    hypercarge_location_cleaned['surroundingChargers'].isnull()
]

# Count the number of rows that meet this condition
count_missing_combination = missing_combination.shape[0]

# Display the result
print("Number of rows where gpsLat, gpsLong, and surroundingChargers are all NaN:", count_missing_combination)

Number of rows where gpsLat, gpsLong, and surroundingChargers are all NaN: 115


We don't know how too fill NaN values for these features. For now, keep them.

In [157]:
# Find features with missing values
missing_values = hypercarge_location_cleaned.isnull().sum()

# Filter to show only columns with missing values
missing_values = missing_values[missing_values > 0]

# Display features with missing values and the count
print(missing_values)

gpsLat                 115
gpsLong                115
surroundingChargers    115
dtype: int64


## Diagnose individual features

In [158]:
# Get the data types of each feature in the dataset
feature_types = hypercarge_location_cleaned.dtypes

# Get the count of unique values for each feature in the dataset
unique_value_counts = {feature: hypercarge_location_cleaned[feature].nunique() for feature in hypercarge_location_cleaned.columns}

# Create a DataFrame to display features, their types, and unique value counts
combined_info = pd.DataFrame({
    'Feature': feature_types.index,
    'Type': feature_types.values,
    'Unique Value Count': [unique_value_counts[feature] for feature in feature_types.index]
})

# Display the combined DataFrame
print(combined_info)

                      Feature     Type  Unique Value Count
0                   chargerId    int64                 374
1                serialNumber   object                 374
2                      sshurl   object                 374
3                      gpsLat  float64                 232
4                     gpsLong  float64                 232
5                numberStacks    int64                   3
6                     chassis   object                   3
7               emergencyStop     bool                   1
8           doorContactSwitch     bool                   2
9         standAloneOrBackend   object                   1
10                   isPublic     bool                   1
11    prohibitSoftwareupdates     bool                   1
12         needsSpecialAccess     bool                   1
13             notServiceable     bool                   1
14              deadManSwitch     bool                   2
15                  hwVersion    int64                  

List of deleted features and reason(s):
1. Technical Specifications not directly influencing customer behaviour: serialNumber, sshurl, outletList, allNotes; status_updatetime
2. Only have 1 value, don't have valuable analysis result: standAloneOrBackend, isPublic, prohibitSoftwareupdates, needsSpecialAccess, notServiceable, hwVersion, isRemoteLocation, status_position, status_isPrivate, excludeFromStatistics, isUtilityExecutionBlocked, creditCardTerminalActive, enableClientWebinterface, hasVollmacht, hasCommissioningProtocol, finishedEol


In [159]:
features_to_delete = [
    'serialNumber',
    'sshurl',
    'outletList',
    'allNotes',
    'standAloneOrBackend',
    'isPublic',
    'prohibitSoftwareupdates',
    'needsSpecialAccess',
    'notServiceable',
    'hwVersion',
    'isRemoteLocation',
    'status_position',
    'status_isPrivate',
    'excludeFromStatistics',
    'isUtilityExecutionBlocked',
    'creditCardTerminalActive',
    'enableClientWebinterface',
    'hasVollmacht',
    'hasCommissioningProtocol',
    'finishedEol',
    'status_updatetime',
    'emergencyStop'
]

# Delete the specified features from the dataset
hypercarge_location_cleaned.drop(columns=features_to_delete, inplace=True)

In [160]:
# Get the data types of each feature in the dataset
feature_types = hypercarge_location_cleaned.dtypes

# Get the count of unique values for each feature in the dataset
unique_value_counts = {feature: hypercarge_location_cleaned[feature].nunique() for feature in hypercarge_location_cleaned.columns}

# Create a DataFrame to display features, their types, and unique value counts
combined_info = pd.DataFrame({
    'Feature': feature_types.index,
    'Type': feature_types.values,
    'Unique Value Count': [unique_value_counts[feature] for feature in feature_types.index]
})

# Display the combined DataFrame
print(combined_info)

                   Feature     Type  Unique Value Count
0                chargerId    int64                 374
1                   gpsLat  float64                 232
2                  gpsLong  float64                 232
3             numberStacks    int64                   3
4                  chassis   object                   3
5        doorContactSwitch     bool                   2
6            deadManSwitch     bool                   2
7            telemetrySave    int64                   3
8         hymaintInstalled     bool                   2
9           lastSignOfLife   object                 142
10        status_errorcode   object                   3
11     status_hycErrorCode   object                   3
12           status_status   object                   5
13            simActivated     bool                   2
14                isActive     bool                   2
15         softwareVersion   object                   9
16           distributorId    int64             

NEXT STEPS
1. There are some chargerId have the same gpsLat and gpsLong?
(take in to consideration missing value for gpsLat and gpsLong)
2. lastSignOfLife: considering create new features LastSignOfLife_Year; LastSignOfLife_Month; LastSignOfLife_Week
3. surroundingChargers: count number of surroundingChargers


In [161]:
# Check for chargerId with the same (gpsLat, gpsLong) values, ignoring NaN values
duplicate_coordinates = hypercarge_location_cleaned[
    hypercarge_location_cleaned['gpsLat'].notnull() & 
    hypercarge_location_cleaned['gpsLong'].notnull()
]

# Identify duplicates based on gpsLat and gpsLong
duplicate_charger_ids = duplicate_coordinates[duplicate_coordinates.duplicated(subset=['gpsLat', 'gpsLong'], keep=False)]

# Select chargerId, gpsLat, and gpsLong for display
duplicate_charger_ids = duplicate_charger_ids[['chargerId', 'gpsLat', 'gpsLong']]

# Display the result
print(duplicate_charger_ids)

     chargerId     gpsLat    gpsLong
10       20793  45.517027   9.147218
28       83765  44.609700   8.661300
36       20805  45.633854   9.600788
40       20791  41.595306  15.470685
48       17424  41.849835  12.732820
60       19522  44.826193  10.380317
85       83753  44.427700  11.601300
124      83768  44.427700  11.601300
136      12615  45.461680   9.276506
143      20804  45.517027   9.147218
149      17815  45.142908   9.634451
188      83756  43.094100  11.780600
194      12616  45.461680   9.276506
220      17422  41.849835  12.732820
235      20807  45.633854   9.600788
244      20806  41.595306  15.470685
254       9922  45.461680   9.276506
259      83772  43.094100  11.780600
260      16941  45.142908   9.634451
268      83759  43.094100  11.780600
279      83767  44.609700   8.661300
351      19527  44.826193  10.380317


So there are some chargerId entries that share the same gpsLat and gpsLong values, while ignoring any entries where these values are NaN
=> If it is not data errors, could ignore

Diagnose gpsLong gpsLat

In [162]:
# Define latitude and longitude boundaries for Italy
lat_min = 35.4892
lat_max = 47.0929
long_min = 6.6272
long_max = 18.9837

# Filter for rows where gpsLat and gpsLong are NOT within the boundaries of Italy
not_italy_coordinates = hypercarge_location_cleaned[~(
    (hypercarge_location_cleaned['gpsLat'].notnull()) & 
    (hypercarge_location_cleaned['gpsLong'].notnull()) & 
    (hypercarge_location_cleaned['gpsLat'] >= lat_min) &
    (hypercarge_location_cleaned['gpsLat'] <= lat_max) &
    (hypercarge_location_cleaned['gpsLong'] >= long_min) &
    (hypercarge_location_cleaned['gpsLong'] <= long_max)
)]

# Select relevant columns for chargerId to delete
not_italy_charger_ids = not_italy_coordinates['chargerId']

Delete all these chargerId?

In [163]:
# Delete these chargerId entries from the original dataset
hypercarge_location_cleaned = hypercarge_location_cleaned[~hypercarge_location_cleaned['chargerId'].isin(not_italy_charger_ids)]

Label Encoding

In [164]:
from sklearn.preprocessing import LabelEncoder
# List of features to apply Label Encoding
features_to_encode = [
    'numberStacks',
    'chassis',
    'doorContactSwitch',
    'deadManSwitch',
    'telemetrySave',
    'hymaintInstalled',
    'status_errorcode',
    'status_hycErrorCode',
    'status_status',
    'simActivated',
    'isActive',
    'softwareVersion',
    'distributorId',
    'corporationId',
    'isFavourite',
    'sessionLiveViewEnabled'
]

# Initialize the LabelEncoder
label_encoder = LabelEncoder()

# Apply Label Encoding to each specified feature
for feature in features_to_encode:
    hypercarge_location_cleaned[feature + '_encoded'] = label_encoder.fit_transform(hypercarge_location_cleaned[feature])

print("Label Encoding applied successfully to specified features.")

Label Encoding applied successfully to specified features.


Create new features

In [165]:
# Convert lastSignOfLife to datetime
hypercarge_location_cleaned['lastSignOfLife'] = pd.to_datetime(hypercarge_location_cleaned['lastSignOfLife'], errors='coerce')

# Create new features for Year, Month, and Week
hypercarge_location_cleaned['LastSignOfLife_Year'] = hypercarge_location_cleaned['lastSignOfLife'].dt.year
hypercarge_location_cleaned['LastSignOfLife_Month'] = hypercarge_location_cleaned['lastSignOfLife'].dt.month
hypercarge_location_cleaned['LastSignOfLife_Week'] = hypercarge_location_cleaned['lastSignOfLife'].dt.isocalendar().week

In [166]:
import ast
# Function to count available surrounding chargers
def count_available_surrounding_chargers(surrounding_chargers_str):
    try:
        # Safely evaluate the string representation to convert to a list of dictionaries
        surrounding_chargers_list = ast.literal_eval(surrounding_chargers_str)
        # Count only the chargers that are available
        return sum(1 for charger in surrounding_chargers_list if charger['status'] == 'Available')
    except (ValueError, SyntaxError):
        return 0  # Return 0 if there is an error in parsing

# Apply the function to the surroundingChargers column
hypercarge_location_cleaned['AvailableSurroundingChargers_Count'] = hypercarge_location_cleaned['surroundingChargers'].apply(count_available_surrounding_chargers)

In [167]:
# Find features with missing values
missing_values = hypercarge_location_cleaned.isnull().sum()

# Filter to show only columns with missing values
missing_values = missing_values[missing_values > 0]

# Display features with missing values and the count
print(missing_values)

Series([], dtype: int64)


In [168]:
# Check the statistics of numeric columns
hypercarge_location_cleaned.describe().T

count         mean           std  \
chargerId                           228.0  20391.54386  19913.838057   
gpsLat                              228.0    43.747977      1.595432   
gpsLong                             228.0    11.702939      2.164869   
numberStacks                        228.0     3.960526      0.301573   
telemetrySave                       228.0      0.04386      0.360869   
distributorId                       228.0   187.135965    107.750707   
corporationId                       228.0   189.872807    108.854697   
numberStacks_encoded                228.0     1.969298      0.217971   
chassis_encoded                     228.0     1.013158      0.175109   
doorContactSwitch_encoded           228.0     0.973684      0.160425   
deadManSwitch_encoded               228.0     0.008772      0.093452   
telemetrySave_encoded               228.0     0.030702      0.196725   
hymaintInstalled_encoded            228.0     0.921053       0.27025   
status_errorcode_encoded            228.0     1.008772       0.16234   
status_hycErrorCode_encoded         228.0     1.122807      0.378738   
status_status_encoded               228.0     0.219298      0.831736   
simActivated_encoded                228.0     0.903509      0.295913   
isActive_encoded                    228.0     0.960526      0.195147   
softwareVersion_encoded             228.0     1.907895      0.873158   
distributorId_encoded               228.0     0.004386      0.066227   
corporationId_encoded               228.0     0.035088      0.206921   
isFavourite_encoded                 228.0     0.008772      0.093452   
sessionLiveViewEnabled_encoded      228.0     0.087719      0.283509   
LastSignOfLife_Year                 228.0  2023.986842      0.114201   
LastSignOfLife_Month                228.0     8.942982      0.488902   
LastSignOfLife_Week                 228.0    35.820175      1.930085   
AvailableSurroundingChargers_Count  228.0    35.846491      15.33727   

                                          min        25%        50%  \
chargerId                              6590.0    9919.75    16936.5   
gpsLat                              40.837697  42.203072  44.326951   
gpsLong                               8.44112   9.600788  11.599933   
numberStacks                              1.0        4.0        4.0   
telemetrySave                             0.0        0.0        0.0   
distributorId                           180.0      180.0      180.0   
corporationId                           180.0      180.0      180.0   
numberStacks_encoded                      0.0        2.0        2.0   
chassis_encoded                           0.0        1.0        1.0   
doorContactSwitch_encoded                 0.0        1.0        1.0   
deadManSwitch_encoded                     0.0        0.0        0.0   
telemetrySave_encoded                     0.0        0.0        0.0   
hymaintInstalled_encoded                  0.0        1.0        1.0   
status_errorcode_encoded                  0.0        1.0        1.0   
status_hycErrorCode_encoded               0.0        1.0        1.0   
status_status_encoded                     0.0        0.0        0.0   
simActivated_encoded                      0.0        1.0        1.0   
isActive_encoded                          0.0        1.0        1.0   
softwareVersion_encoded                   0.0        2.0        2.0   
distributorId_encoded                     0.0        0.0        0.0   
corporationId_encoded                     0.0        0.0        0.0   
isFavourite_encoded                       0.0        0.0        0.0   
sessionLiveViewEnabled_encoded            0.0        0.0        0.0   
LastSignOfLife_Year                    2023.0     2024.0     2024.0   
LastSignOfLife_Month                      4.0        9.0        9.0   
LastSignOfLife_Week                      17.0       36.0       36.0   
AvailableSurroundingChargers_Count        3.0       22.0       36.5   

                              

Visualisations

In [171]:
import folium
from IPython.display import display
from folium.plugins import HeatMap

# Set the center of the map to Italy
italy_lat = 41.8719
italy_long = 12.5674
charger_map = folium.Map(location=[italy_lat, italy_long], zoom_start=6)  # Adjust zoom level as needed

# Prepare data for the heatmap
heat_data = [[row['gpsLat'], row['gpsLong']] for _, row in hypercarge_location_cleaned.iterrows()]

# Create a HeatMap layer
HeatMap(heat_data, radius=15).add_to(charger_map)

# Display the heatmap in the notebook
display(charger_map)